runtime/doc/eval.txt

```diff
@@ -1378,6 +1378,51 @@ to be doubled.  These two commands are equivalent: >
 	if a =~ '\s*'
 
 
+template-string					*template-string* *E1000*
+---------------
+
+$"string"	template string			*expr-$quote*
+$'string'	template literal string		*expr-$'*
+
+Note that either single quotes or double quotes are used.
+
+These strings support 'string-interpolations'. That embeds any expressions by
+'${}' into a string. For example
+>
+	" Embedding literals
+	echo $'I have ${10}'          " I have 10
+	echo $'I ${{"love": 1000}}'   " I {'love': 1000}
+<
+These are same as
+>
+	echo 'I have ' . string(10)
+	echo 'I ' . string({"love": 1000})
+<
+'string-interpolations' helps you to write complex string. Other rules:
+>
+	" Only strings or literal strings doesn't apply string()
+	echo $'I am a ${"vim"}'   " I am a vim
+	                          " not: I am a 'vim'
+	echo $"I am a ${'vim'}"   " I am a vim
+
+	" Only unscoped variables can omit {}
+	let x = 10
+	echo $'$x'  " 10
+
+	" In template-string, you can write '$' as $$
+	echo $'$$'  " $
+
+	" Escaping in template (non-literal) string
+	echo $"${10} \\ \""   " 10 \ "
+<
+Exceptions:
+>
+	echo $'${var'  " E1000: Unterminated template literal: $'${var'
+	echo $'$10'    " E1000: Illegal template literal: $'$10'
+	echo $'... $ ...'  " E1000: Missing variable name on a $: $'... $ ...'
+	echo $'---${}---'  " E1000: Empty template literal: $'---${}---'
```

Comment:
I think that different error messages should use different error number.

Reply:
mmm, I want to use `E1000`.
Because Now `E999` is the maximum number of exceptions, and I love a number `1000` x(

But I cannot refuse this if this is really needed x)

Reply:
Just FYI, there are still many error numbers available before E1000.
https://github.com/vim/vim/blob/e8a7dfedfc8ea5c376c7912cb27a9405e4b8c972/runtime/doc/todo.txt#L132-L135
E.g. E450 to E453 might be able to be used for these 4 error messages.

Reply:
Understood, I'll use these number before E1000 😎

Now I'm using only E1000 for new behaviors.
Should I use different numbers (e.g. E450 to E453) for these?

Reply:
Bram will decide the error numbers when merging. You can use arbitrary numbers while developing.

> Should I use different numbers (e.g. E450 to E453) for these?

Not sure.
At least, I think it's better to use different numbers. This has a merit that you can catch the different errors in your tests.

Reply:
Sorry, I understood.
I entrust about E1000 to Bram.
Thanks a lot!

Also the merit seems good.
I'll use these different numbers 👍️

Reply:
I fixed :+1:

```vim
  try
    let _ = $'$10'
    call assert_report('Should throw an exception.')
  catch
    call assert_exception('E450:')
  endtry

  try
    let _ = $'---$---'
    call assert_report('Should throw an exception.')
  catch
    call assert_exception('E450:')
  endtry

  try
    let _ = $'${10'
    call assert_report('Should throw an exception.')
  catch
    call assert_exception('E451:')
  endtry

  try
    let _ = $'---${}---'
    call assert_report('Should throw an exception.')
  catch
    call assert_exception('E452:')
  endtry
```